sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -176,10 +176,16 @@ abstract class UnaryNode extends LogicalPlan {
         allConstraints += EqualNullSafe(a.toAttribute, l)
       case a @ Alias(e, _) =>
         // For every alias in `projectList`, replace the reference in constraints by its attribute.
-        allConstraints ++= allConstraints.map(_ transform {
-          case expr: Expression if expr.semanticEquals(e) =>
-            a.toAttribute
-        })
+        allConstraints ++= allConstraints.map {
+          case e @ EqualNullSafe(l, _: AttributeReference)
+            if !l.isInstanceOf[AttributeReference] => e
```

Comment:
what's the rationale behind this change?

Reply:
For example:
```scala
import org.apache.spark.sql.catalyst.plans.logical.Project
spark.sql("CREATE TABLE IF NOT EXISTS spark_29606(a int, b int, c int) USING parquet")
spark.sql("SELECT a as a1, b as b1, c as c1, abc as abc1 FROM (SELECT a, b, c, a + b + c as abc FROM spark_29606) t")
  .queryExecution.analyzed.asInstanceOf[Project].validConstraints.toSeq.sortBy(_.toString).foreach(println)
```
`child.constraints` has a  constraint: `(((a#5 + b#6) + c#7) <=> abc#0)`:


![image](https://user-images.githubusercontent.com/5399861/69068511-dc27fb00-0a5f-11ea-98ab-dd0c52c27529.png)


**Before this PR**.  We will replace the reference in constraints by its attribute. It will generate a lot of constraints. It seems that these constraints are useless:
![image](https://user-images.githubusercontent.com/5399861/69068611-0a0d3f80-0a60-11ea-98ca-4b23839c29cf.png)


**After this PR**: We avoid generating these useless constraints:
![image](https://user-images.githubusercontent.com/5399861/69069075-d0890400-0a60-11ea-80e5-e54d162fb702.png)

Reply:
This is just an example. Why can we ignore `EqualNullSafe` between attr and non-attr? Are they guaranteed to produce useless constraints?

Reply:
`validConstraints` is used here:
https://github.com/apache/spark/blob/15de6d05008cf8054fe9c2246ee80f607f9f33e0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala#L30-L43

1. These generated constraint are not the `subsetOf(outputSet)`.
2. These generated constraint will not affect `inferAdditionalConstraints` and `constructIsNotNullConstraints `.